apps/aecore/src/aec_accounts.erl

```diff
@@ -73,35 +108,85 @@ spend(#account{balance = Balance0} = Account0, Amount, Nonce) ->
 spend_without_nonce_bump(#account{balance = Balance0} = Account0, Amount) ->
     {ok, Account0#account{balance = Balance0 - Amount}}.
 
+-spec type(account()) -> basic | generalized.
+type(#account{ ga_contract = undefined }) -> basic;
+type(#account{ ga_contract = C })         -> contract = aeser_id:specialize_type(C), generalized.
+
+is_legal_at_height(Account, Height) ->
+    case type(Account) of
+        basic       -> true;
+        generalized -> aec_hard_forks:protocol_effective_at_height(Height) >= ?FORTUNA_PROTOCOL_VSN
+    end.
+
 -spec serialize(account()) -> deterministic_account_binary_with_pubkey().
 serialize(Account) ->
+    case type(Account) of
+        basic -> serialize_basic(Account);
+        generalized -> serialize_generalized(Account)
+    end.
+
+serialize_basic(Account) ->
     aeser_chain_objects:serialize(
-      ?ACCOUNT_TYPE, ?ACCOUNT_VSN,
-      serialization_template(?ACCOUNT_VSN),
+      ?ACCOUNT_TYPE, ?ACCOUNT_VSN_1,
+      serialization_template(?ACCOUNT_VSN_1),
       [ {nonce, nonce(Account)}
       , {balance, balance(Account)}
       ]).
 
+serialize_generalized(Account) ->
+    aeser_chain_objects:serialize(
+      ?ACCOUNT_TYPE, ?ACCOUNT_VSN_2,
+      serialization_template(?ACCOUNT_VSN_2),
+      [ {flags, flags(Account)}
+      , {nonce, nonce(Account)}
+      , {balance, balance(Account)}
+      , {ga_contract, ga_contract(Account)}
+      , {ga_auth_fun, ga_auth_fun(Account)}
+      ]).
+
 -spec deserialize(aec_keys:pubkey(), binary()) -> account().
 deserialize(Pubkey, SerializedAccount) ->
-    [ {nonce, Nonce}
-    , {balance, Balance}
-    ] = aeser_chain_objects:deserialize(
-          ?ACCOUNT_TYPE,
-          ?ACCOUNT_VSN,
-          serialization_template(?ACCOUNT_VSN),
-          SerializedAccount),
-    #account{ id = aeser_id:create(account, Pubkey)
-            , balance = Balance
-            , nonce = Nonce
-            }.
-
-serialization_template(?ACCOUNT_VSN) ->
+    case aeser_chain_objects:deserialize_type_and_vsn(SerializedAccount) of
+        {?ACCOUNT_TYPE = Type, ?ACCOUNT_VSN_1 = Vsn, _Rest} ->
+            [ {nonce, Nonce}
+            , {balance, Balance}
+            ] = aeser_chain_objects:deserialize(
+                    Type, Vsn, serialization_template(Vsn), SerializedAccount),
+            #account{ id = aeser_id:create(account, Pubkey)
+                    , balance = Balance
+                    , nonce = Nonce
+                    };
+        {?ACCOUNT_TYPE = Type, ?ACCOUNT_VSN_2 = Vsn, _Rest} ->
+            [ {flags, Flags}
+            , {nonce, Nonce}
+            , {balance, Balance}
+            , {ga_contract, GAContract}
+            , {ga_auth_fun, GAAuthFun}
+            ] = aeser_chain_objects:deserialize(
+                    Type, Vsn, serialization_template(Vsn), SerializedAccount),
+            #account{ id = aeser_id:create(account, Pubkey)
```

Comment:
It might be prudent to assert that this is indeed a generalized account. To prevent leaking in basic accounts in version 2 in the deserialization (from a malicious party).

Reply:
Noted: I will strengthen the checks - not necessarily in this PR.